internal/importer/xml.go

```diff
@@ -54,1009 +53,504 @@ func init() {
-	root := &book.Node{
-		Kind: book.MapNode,
-		Children: []*book.Node{
-			{
-				Kind:  book.MapNode,
-				Name:  book.KeywordSheet,
-				Value: book.MetasheetName,
-			},
-		},
-	}
-	// sheetName -> {colName -> val}
-	metasheetMap := make(map[string]map[string]string)
-	for n := tableauNode.FirstChild; n != nil; n = n.NextSibling {
-		if n.Type != xmlquery.ElementNode {
-			continue
-		}
-		var children []*book.Node
-		sheetMap := make(map[string]string)
-		var sheetName string
-		for _, attr := range n.Attr {
-			name := attr.Name.Local
-			value := attr.Value
-			children = append(children, &book.Node{
-				Name:  name,
-				Value: value,
-			})
-			sheetMap[name] = value
-
-			if name == "Sheet" {
-				sheetName = value
-			}
-		}
-		if sheetName == "" {
-			return metasheetMap, nil, errors.Errorf("field `Sheet` not specified in metasheet @TABLEAU")
-		}
-		metasheetMap[sheetName] = sheetMap
-		sheetNode := &book.Node{
-			Kind:     book.MapNode,
-			Name:     sheetName,
-			Children: children,
+func parseXMLSheet(doc *xmldom.Document, mode ImporterMode) (*book.Sheet, error) {
+	name := doc.Root.Name
+	if mode == Protogen {
+		if name == atTableauDisplacement {
+			return parseXMLMetaSheet(doc)
 		}
-		root.Children = append(root.Children, sheetNode)
-	}
-	doc := &book.Node{
-		Kind:     book.DocumentNode,
-		Name:     book.MetasheetName,
-		Children: []*book.Node{root},
-	}
-	sheet := book.NewDocumentSheet(book.MetasheetName, doc)
-	return metasheetMap, sheet, nil
-}
-
-// addMetaNodeAttr adds an attribute to MetaNode AttrMap if not exists and otherwise replaces the attribute value
-func addMetaNodeAttr(attrMap *tableaupb.XMLNode_AttrMap, name, val string) {
-	if idx, ok := attrMap.Map[name]; !ok {
-		attrMap.Map[name] = int32(len(attrMap.List))
-		attrMap.List = append(attrMap.List, &tableaupb.XMLNode_AttrMap_Attr{
-			Name:  name,
-			Value: val,
-		})
-	} else {
-		// replace attribute value by metaSheet
-		metaAttr := attrMap.List[idx]
-		metaAttr.Value = val
+		name = "@" + name
 	}
-}
 
-// addDataNodeAttr adds an attribute to DataNode AttrMap
-func addDataNodeAttr(metaMap, dataMap *tableaupb.XMLNode_AttrMap, name, val string) {
-	if _, ok := metaMap.Map[name]; !ok {
-		metaMap.Map[name] = int32(len(metaMap.List))
-		metaMap.List = append(metaMap.List, &tableaupb.XMLNode_AttrMap_Attr{
-			Name:  name,
-			Value: inferType(val),
-		})
-	}
-	dataMap.Map[name] = int32(len(dataMap.List))
-	dataMap.List = append(dataMap.List, &tableaupb.XMLNode_AttrMap_Attr{
-		Name:  name,
-		Value: val,
+	bnode := &book.Node{}
+	bnode.Kind = book.MapNode
+	bnode.Children = append(bnode.Children, &book.Node{
+		Name:  book.SheetKey,
+		Value: name,
 	})
-}
 
-// hasChild check if xml node has any child element node
-func hasChild(n *xmlquery.Node) bool {
-	for n := n.FirstChild; n != nil; n = n.NextSibling {
-		if n.Type == xmlquery.ElementNode {
-			return true
-		}
+	rootNode := &book.Node{}
+	parseXMLNode(doc.Root, rootNode, mode)

```

Comment:
Must check returned error. 

Please also fix it in other postitions.

Reply:
done